File: Test/BigInteger/modinv.pas
var b := BigInteger(3);
PrintLn(b.ModInv(BigInteger(11)));

b := BigInteger(40);
PrintLn(b.ModInv(BigInteger(1)));

b := BigInteger(42);
PrintLn(b.ModInv(BigInteger(2017)));

b := BigInteger(52);
PrintLn(b.ModInv(BigInteger(-217)));

b := BigInteger(-486);
PrintLn(b.ModInv(BigInteger(217)));

b := BigInteger(154);
PrintLn(b.ModInv(BigInteger(3311)));

File: Libraries/JSCodeGen/rtl/ModInv.js
function ModInv(v, n) {
    let t = 0n, nT = 1n, r = n, nR = v < 0 ? -v : v;
    while (nR != 0) {
        let q = r / nR, pT = t, pR = r;
        t = nT;
        r = nR;
        nT = pT - q*nT;
        nR = pR - q*nR;
    }
    if (r != 1 && r != -1) return 0n;
    if (t < 0) t += n;
    return v < 0 ? n - t : t;
}
